ssl: Add support for {active,N}

essen opened the pull request:
<s>There's probably missing stuff for DTLS, and the documentation is also missing, but this is a good time to get a first review. The added tests pass and check the option with listen, server and client sockets. I'll probably need some guidance about DTLS.</s>

All functionality is there and tests pass! Documentation and possibly some refactoring remains.

IngelaAndin commented:
We have not had time to review the code yet but the new test is failing quite a lot.
Also other tests are failing but it not clear yet if that is related or not. More feedback to come
here are some examples of how the test fails at the moment.
 
```
** CT Error Notification 2019-01-07 07:08:53.319 ***
ssl_basic_SUITE:'-active_n/1-fun-1-' failed on line 2090
Reason: error

*** CT Error Notification 2019-01-07 07:04:05.217 ***
ssl_basic_SUITE:active_n_common failed on line 2134
Reason: error

```

essen commented:
How did you get these line numbers? 2090 is not inside a fun here. Could you point in the diff where it occurs? Thanks.

IngelaAndin commented:
I built ssl on your PR branch and then I get the following:

```
*** CT Error Notification 2019-01-07 15:55:07.207 ***
ssl_basic_SUITE:active_n_common failed on line 2099
Reason: error

2093: active_n_common(S, N) ->
 2094:     ok = ssl:setopts(S, [{active,-N}]),
 2095:     receive
 2096:         {ssl_passive, S} -> ok
 2097:     after
 2098:         1000 ->
 2099:             error({error,ssl_passive_failure})
 2100:     end,
 2101:     [{active,false}] = ok(ssl:getopts(S, [active])),
 2102:     ok = ssl:setopts(S, [{active,0}]),
 2103:     receive
 2104:         {ssl_passive, S} -> ok
 2105:     after
 2106:         1000 ->
 2107:             error({error,ssl_passive_failure})
 2108:     end,
 2109:     ok = ssl:setopts(S, [{active,32767}]),
 2110:     {error,{options,_}} = ssl:setopts(S, [{active,1}]),
 2111:     {error,{options,_}} = ssl:setopts(S, [{active,-32769}]),
 2112:     ok = ssl:setopts(S, [{active,-32768}]),
 2113:     receive
 2114:         {ssl_passive, S} -> ok
 2115:     after
 2116:         1000 ->
 2117:             error({error,ssl_passive_failure})
 2118:     end,
 2119:     [{active,false}] = ok(ssl:getopts(S, [active])),
 2120:     ok = ssl:setopts(S, [{active,N}]),
 2121:     ok = ssl:setopts(S, [{active,true}]),
 2122:     [{active,true}] = ok(ssl:getopts(S, [active])),
 2123:     receive
 2124:         _ -> error({error,active_n})
 2125:     after
 2126:         0 ->
 2127:             ok
 2128:     end,
 2129:     ok = ssl:setopts(S, [{active,N}]),
 2130:     ok = ssl:setopts(S, [{active,once}]),
 2131:     [{active,once}] = ok(ssl:getopts(S, [active])),
 2132:     receive
 2133:         _ -> error({error,active_n})
 2134:     after
 2135:         0 ->
 2136:             ok
 2137:     end,
 2138:     {error,{options,_}} = ssl:setopts(S, [{active,32768}]),
 2139:     ok = ssl:setopts(S, [{active,false}]),
 2140:     [{active,false}] = ok(ssl:getopts(S, [active])),
 2141:     ok.

```

essen commented:
I have pushed a fix for the sslv3 and tls1 cases. I have not yet looked into implementing DTLS but I will soon.

IngelaAndin commented:
As the option is emulated and never set on the real socket  the handling ought to be quite similar.

essen commented:
I've pushed the commit for DTLS. The tests pass and the functionality should be complete. Documentation will still need to be updated.

The code looks very similar between tls/dtls but it's difficult to abstract because of small differences between tls_socket/dtls_socket. Open to suggestions.

essen commented:
And one more push to remove some unnecessary code and hopefully fix Dialyzer.

IngelaAndin commented:
The below is no longer true form the user perspective. An optimization will add those two application data  TLS-records into one application data delivery. Hopefully I can merge my branch in the beginning of next week so that you can see it to. Otherwise test seems to be doing  good now. I hope my team will have some time to look closer at the code soon.

```
%% For TLSv1 and SSLv3 we receive the first byte as a separate message.
 2045:     %% Therefore we need twice the number of active messages.
 2046:     {ok, Infos} = ssl:connection_information(C, [protocol]),
 2047:     Multiplier = case Infos of
 2048:         [{_, sslv3}] -> 2;
 2049:         [{_, tlsv1}] -> 2;
```

essen commented:
Sounds good!

garazdawi commented on lib/ssl/src/dtls_socket.erl:
```diff
@@ -81,15 +83,43 @@ socket(Pids, Transport, Socket, ConnectionCb) ->
+                Atom when is_atom(Atom), N =< 0 ->
+                    self() ! {ssl_passive, Socket};
+                %% The result of the addition is out of bound.
+                A when is_integer(A), A + N < -32768 ->
```
Can this happen? Wouldn't A have to be < 0?

essen replied:
Yes but no test covers that before, the `N =< 0` in the clause above only applies if `Active` is `true | false | once`. This clause is when `Active` is an integer.

garazdawi replied:
So `{ok, #socket_options{active = Active}, _} = dtls_packet_demux:get_all_opts(ListenPid),` can return an `Active` that is -1? When does that happen?

essen replied:
Confused the variable. You're right, that clause is probably not necessary. I will remove it and run the tests again.

essen replied:
Removed the clause in 0f6dc6862efcaa1c0a241ddd81d805e65e7aded7.

IngelaAndin commented:
Due to test problems unrelated to the ssl application I can not merge today. But hopefully tomorrow we I can merge so that you can remove multiplier workaround in the test case.  When it comes to the documentation it probably does not need to be so extensive and mostly refer to inet. As it happens I am just in the process of reworking the documentation to use type specs so preferably we will postpone the documentation until I am done with that.

essen commented:
I will try removing the unreachable clause tomorrow and I will then wait for the documentation.

Note that the documentation in inet doesn't mention `tcp_passive`, it's mentioned in `gen_tcp:connect` but that doesn't look like the right place. Might be worth revisiting at the same time.

IngelaAndin commented:
I have now merge my optimization sha:  6ce69af1ae46b6b09c5a7ecd8b1baa1dcde77e24  that
should render multiplier workaround in test unnecessary.

essen commented:
I've rebased and removed the multiplier workaround.

IngelaAndin commented:
I have now pushed my documentation branch.  Also after fixing a DTLS error handling bug the active_n test are running flawlessly on all platforms.  Look for activ_msgs for good place to add documentation.

essen commented:
I've rebased, made the code changes into one commit and added a second commit for the documentation changes. Please review.

Would be great to get this in 21.3!

IngelaAndin commented on lib/ssl/src/dtls_socket.erl:
```diff
@@ -81,15 +83,42 @@ socket(Pids, Transport, Socket, ConnectionCb) ->
+            throw(einval);
+        {_, N} when is_integer(N), N > 32767 ->
+            throw(einval);
+        {_, N} when is_integer(N) ->
```
The check of active N value is done in several places in a similar way, I was thinking it could be nice with some helper function that preforms the actual check.

essen replied:
It's a little difficult to have a common `check_active_n` because we have 3 scenarios where it is called: DTLS sockets, TLS sockets and TLS listening sockets. The code path for each of these is different. In the implementation of this PR the latter 2 share a common `check_active_n` while DTLS has its own. I'm not seeing an obvious way to share the code between the 3 currently.

Of course the two clauses checking the upper and lower bounds are common but we don't gain much (if anything) from refactoring that.

bmk commented on lib/ssl/src/dtls_packet_demux.erl:
```diff
@@ -297,6 +297,18 @@ do_set_emulated_opts([], Opts) ->
     Opts;
 do_set_emulated_opts([{mode, Value} | Rest], Opts) ->
     do_set_emulated_opts(Rest,  Opts#socket_options{mode = Value}); 
+do_set_emulated_opts([{active, N0} | Rest], Opts=#socket_options{active = Active}) when is_integer(N0) ->
```
I noticed this construction several times. Is it possible to combine into one (utility) function?

/BMK

essen replied:
I have added a common function for this.

essen commented:
I'll try to refactor a bit.

IngelaAndin commented:
I tried to put your branch in testing again, but I am afraid that my branch that shrinks the outer state record,that I merged today creates some problems for your branch. Could you please rebase on latest maint.

IngelaAndin commented:
That is problems are not seen as merge conflicts but rather as compiler errors after merging as state data has been moved around a lot.

essen commented:
I have fixed the issue.

This also made one of the tests fail. The fix is in the most recent commit. Please check it. Not sure it is necessary to update `socket_options` in other clauses.

IngelaAndin commented:
I think the code is looking good :) And I do think that it will be sufficient for now to update the socket_options in the place that you have. However, it might be interesting to handle this
in the new dist_read_application_data @RaimoNiskanen if there could be a benefit in making it  
{active, N} instead of active?  Just saying we should think about it.

IngelaAndin commented:
We are currently running our own branch that includes you commits to speed up testing so it can make OTP-21.3, as this time we got new collisions with our changes.

essen commented:
Glad to hear, especially considering I am fairly busy at the moment.

essen commented:
All good?

IngelaAndin commented:
Yes it is looking good, we will merge our version of the branch today. I was away on "sports holiday"  (Swedish kids has a week of school to do winter sports) last week there of the delay.

essen commented:
Swedish kids are lucky. :-) Cheers

RaimoNiskanen commented:
I did a local repository merge to resolve the merge conflict.
